### ISSUE TITLE
Failed to instantiate provider "registry.terraform.io/hashicorp/time"

### ISSUE THREAD

**Issue by instagendleg** [hashicorp/terraform#27535]
### Terraform Version

```
Terraform v0.14.4
Go runtime v1.15.6
```

### Terraform Configuration Files

n/a

### Debug Output

https://gist.github.com/instagendleg/7a76a86171011529f01d611202110352

### Crash Output

```log
Failed to instantiate provider "registry.terraform.io/hashicorp/time" to obtain schema: unknown provider "registry.terraform.io/hashicorp/time
```

### Expected Behavior

The provider should be downloaded by the `init` before the `plan`.

### Actual Behavior

The provider `registry.terraform.io/hashicorp/time` is not downloaded by the `init` command. It crashes the `plan` action.

### Steps to Reproduce

1. `terraform init`
2. `terraform plan`

### Additional Context

n/a

### References

n/a

**Comment by instagendleg** [comment]
Maybe be related to #27516.

**Comment by instagendleg** [comment]
I just tested by hard-coding the provider dependency in my `main.tf` root module. This fix works, but it is irrelevant because I do not use any function/module from this provider.

**Added code:**

```hcl
terraform {
  [...]

  required_providers {
    time = {
      version = "~> 0.6"
    }
  }
}
```

**Comment by alisdair** [comment]
Thanks for reporting this. Unfortunately we need more details to reproduce.

- Can you share a configuration that results in this error?
- What is the output of `terraform providers`?

**Comment by ljfraney** [comment]
I ran into the same problem today with 0.13.5.

**Steps to Reproduce:**
1. Add a `time_sleep` resource to your configuration.
2. Run `terraform plan`. This should download the time provider.
3. Run `terraform apply`. Now the `time_sleep` resource is referenced in tfstate.
4. Remove the `time_sleep` resource from your configuration.
5. Run `terraform init`. You no longer have the time provider because it isn't referenced in the configuration.
6. Run `terraform plan`. It fails with the error "Failed to instantiate provider "registry.terraform.io/hashicorp/time"", presumably because the `time_sleep` resource still exists in tfstate.

**Workaround:**
1. Add the time provider in required_providers explicitly as @instagendleg shows above.
2. Run `terraform plan`. This should download the time provider even though it isn't referenced in the configuration.
3. Run `terraform apply`. The `time_sleep` resource is destroyed and no longer referenced in tfstate.
4. Remove the time provider from required_providers.

**Comment by alisdair** [comment]
Thanks for that, @ljfraney. Unfortunately I'm still unable to reproduce the problem, and I can't understand how it could be happening. Here's what I've tried, with both Terraform 0.13.5 and 0.14.4:

1. Create this configuration:

    ```hcl
    resource "null_resource" "one" {}

    resource "time_sleep" "wait" {
      depends_on = [null_resource.one]
      create_duration = "1s"
    }

    resource "null_resource" "two" {
      depends_on = [time_sleep.wait]
    }
    ```
2. `terraform init`
3. `terraform apply -auto-approve`
4.  Remove the `time_sleep` resource and the `depends_on` reference to it
5. `terraform init` (still shows that the `time` provider is required)
6. `terraform providers` (shows that the `time` provider is required by resources in state)
7. `terraform plan` (succeeds)

I've also tried wiping the `.terraform` directory after step 4, which doesn't change anything.

Are you able to modify these steps to reproduce the issue?
